src/test/java/com/fasterxml/jackson/core/util/ReadConstrainedTextBufferTest.java

```diff
@@ -33,22 +46,64 @@ public void appendString() throws Exception {
 
     @Test
     public void appendSingle() throws Exception {
-        TextBuffer constrained = makeConstrainedBuffer(SEGMENT_SIZE);
+        TextBuffer constrained = makeConstrainedBuffer(SEGMENT_SIZE, false);
         char[] chars = new char[SEGMENT_SIZE];
         Arrays.fill(chars, 'A');
         constrained.append(chars, 0, SEGMENT_SIZE);
         Assertions.assertThrows(IOException.class, () -> constrained.append('x'));
     }
 
-    private static TextBuffer makeConstrainedBuffer(int maxStringLen) {
+    private static TextBuffer makeConstrainedBuffer(int maxStringLen, boolean pooledBufferRecycler) {
         StreamReadConstraints constraints = StreamReadConstraints.builder()
                 .maxStringLength(maxStringLen)
                 .build();
         IOContext ioContext = new IOContext(
                 constraints,
                 StreamWriteConstraints.defaults(),
-                new BufferRecycler(),
+                pooledBufferRecycler ? BufferRecyclers.getBufferRecycler() : new BufferRecycler(),
                 ContentReference.rawReference("N/A"), true);
         return ioContext.constructReadConstrainedTextBuffer();
     }
+
+    @Test
+    public void appendStringUsingExistingBufferRecycler() throws Exception {
+        int size = 6000;
+        AsyncReaderWrapper reader = createReader(size);
+
+        // This statement will cause the release of a large buffer in the BufferRecycler.CHAR_TEXT_BUFFER bucket
+        reader.nextToken();
+
+        reader.close();
+
+        // Since it is running on the same thread the ThreadLocal based pooling will make this to reuse the same BufferRecycler
+        TextBuffer constrained = makeConstrainedBuffer(SEGMENT_SIZE, true);
+
+        char[] chars = new char[SEGMENT_SIZE];
+        Arrays.fill(chars, 'A');
+        constrained.append(new String(chars), 0, SEGMENT_SIZE);
+
+        // This should throw an exception but doesn't since the constraint is calculated on the size of the formerly released buffer
+        Assertions.assertThrows(IOException.class, () -> constrained.append(new String(chars), 0, SEGMENT_SIZE));
```

Comment:
I'm not worried about this yet. The `append` check is currently not guaranteed to catch when the contents are too long. We know this already. The code was written this way to try to avoid doing an expensive exact size check on all appends.

We have a check in the `contentsAsString` so the StreamConstraintsException will be thrown at that point

Try changing your test to see this in action.

```
        // This should throw an exception but doesn't since the constraint is calculated on the size of the formerly released buffer
        //Assertions.assertThrows(IOException.class, () -> constrained.append(new String(chars), 0, SEGMENT_SIZE));

        constrained.append(new String(chars), 0, SEGMENT_SIZE);
        constrained.contentsAsString(); // fails here
```

To reiterate, this is a deliberate decision to try to maintain performance for the usual scenario where the appends are safe while trying to in some or most cases catch the fact the string is too large at an early stage.

Since buffers are being reused - we are not using much extra memory here.

Reply:
this appears in our javadocs:

This limit is not exact, the limit is applied when we increase internal buffer sizes and an exception will happen at sizes greater than this limit.

Reply:
@pjfanning ok, thanks for explanation, I also confirm that calling `contentsAsString` throws an IOException as expected. I will close this pull request and the corresponding bug report and modify the other tests in `ReadConstrainedTextBufferTest` so that they will succeed regardless if its using a brand new `BufferRecycler` or one taken from the pool.

Reply:
* please don't modify the existing tests - if they can't use buffer-recyler, so be it
* if we have to add separate tests that use buffer-recyclers but that behave a little bit differently, that is fine